Add guides introducing XOD Cloud Feeds (#158)
xodio/xod-docs: docs/guide/multiple-time-series/README.md

---
title: Multiple time series and bundled data
version: 1.0.0
---

# Multiple time series and bundled data

> To follow the guide you have to know [how to send data to XOD Cloud Feeds](../getting-started/).

Reviewer: Wrong link

> XOD Cloud feeds service provides a possibility to send the bundled data so that it will come as one portion of the data with the same timestamp. It gives the option to see all values on the same table and draw a multiple time series on the same chart.

Reviewer: ```suggestion
The XOD Cloud Feeds service provides a feature to send bundled data so that it comes as an atomic portion with a single timestamp. It gives the option to see all the values on the same table and draw multiple time series on the same chart.
```

> ![Example of a chart with time series](./chart.png)

Reviewer: ```suggestion
For example, you can track temperature inside and outside of your house at the same time.

![Example of a chart with time series](./chart.png)
```

> Note, that data may contain values of different types. For example, imagine that you want to log temperature changes and whether the window is open at any given time. For this case you can bundle the both values with keys `temperature` and `window-opened` and then see it in the table like this:

Reviewer: ```suggestion
Also, that data may contain values of different types. Suppose you want to log temperature changes and whether the window is open at any given time. In this case, you can bundle both values with keys `temperature` and `window-opened` and then see them in a table like this:
```

<table class="ui celled table">
  <thead>
    <tr>
      <th>Date</th>
      <th>temperature</th>
      <th>window-opened</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class="collapsing">2019-11-13 16:48:30</td>
      <td colspan="1" class="number">20.9</td>
      <td colspan="1" class="boolean">true</td>
    </tr>
    <tr>
      <td class="collapsing">2019-11-13 16:47:23</td>
      <td colspan="1" class="number">21.6</td>
      <td colspan="1" class="boolean">true</td>
    </tr>
    <tr>
      <td class="collapsing">2019-11-13 16:46:26</td>
      <td colspan="1" class="number">22.1</td>
      <td colspan="1" class="boolean">true</td>
    </tr>
    <tr>
      <td class="collapsing">2019-11-13 16:45:23</td>
      <td colspan="1" class="number">22.5</td>
      <td colspan="1" class="boolean">false</td>
    </tr>
    <tr>
      <td class="collapsing">2019-11-13 16:44:22</td>
      <td colspan="1" class="number">22.4</td>
      <td colspan="1" class="boolean">false</td>
    </tr>
    <tr>
      <td class="collapsing">2019-11-13 16:43:27</td>
      <td colspan="1" class="number">22.2</td>
      <td colspan="1" class="boolean">false</td>
    </tr>
  </tbody>
</table>

> To store the bundled data, you may use the same node as usual `xod-cloud/messaging/cloud-store,` but you have to send values in JSON format. If the string contains a valid JSON, it will be parsed and stored as a bundled data.

Reviewer: ```suggestion
To store the bundled data, use the same node, as usual, that is `xod-cloud/messaging/cloud-store`. What changes is that you have to send values in the so-called JSON format. If the string passed to `IN` contains a valid JSON, it will be parsed and stored as bundled data.
```

:exclamation: a hyperlink to the node docs is desired

## JSON format

> JSON is an open-standard file format that uses a human-readable text to transmit data objects consisting of attribute-value pairs or array. In the case of sending array, you'll get the same table as shown above, but the index of values will be the key for each value.

Reviewer: ```suggestion
JSON is an open-standard format that uses a human-readable text to transmit data objects consisting of attribute-value pairs or arrays. If you send array instead of an object you will get the same result as above but the columns will have names `0` and `1`.
```

> Here is an example of sending a bundled data: `{"num": 23.5, "bool": true, "str": "Hello World"}`.

Reviewer: The sequence is messed up here. Suggest change the order to the following:

- JSON is an open-standard format that uses a human-readable text to transmit data objects consisting of attribute-value pairs or arrays.
- Here is an example of bundled data portion: `use an example which corresponds to a row of the example above`
- If you send array instead of an object you will get the same result as above but the columns will have names `0` and `1`. Example: `the same values but as an array`

> NOTE: A JSON format is picky about quotes, forgotten or excessive commas, and so on. So if you see a JSON-like string in your feed, it means there are some mistakes.

Reviewer: Please surround with note tags

> ## Bundle the data

Reviewer: ```suggestion
## Bundle data
```

> Let's make a patch that will send a value of the potentiometer, the button state, and string from a `tweak-string` node on any change of them.

Reviewer: ```suggestion
Let's make a patch that sends a value of a potentiometer, a button state, and a string from the `tweak-string` node whenever any of them changes.
```

![Logging multiple time series in XOD](./multiple-time-series.patch.png)

> First of all, pay attention to the `concat` node that we're using to format JSON. Do not forget to enquote each key, but link values as is, except strings, because we have to force enquoting of the string for JSON format.

Reviewer: Too complex. Suggest splitting to UL:

- Enquote each key
- Do not forget `:` and `,`
- Enquote strings
- Do not enquote numbers and booleans

> Then let's examine a pattern of changing: every value that changes produce a pulse, which combined through the `any` node. Note that a potentiometer can provide a noise so that it may send redundant data to the feeds. In this case, you can add a `debounce` node under the `pot.` Also, you can add a `throttle` node to limit the rate of sending data, but you should be ready to miss some changes.

Reviewer: - `any` node? Looks obsolete.
- what the result? I’m curious. An example table would be nice
- The info about throttling is useful but doesn’t really belong here. At the very least extract it to a separate section (Throttling updates), provide a patch with `throttle`, hyperlink to the node
